document the custom render option (#1711) (#1811)
observablehq/plot: docs/features/marks.md

The **sort** option, when not specified as a channel value (such as a field name or an accessor function), can also be used to [impute ordinal scale domains](./scales.md#sort-mark-option).

> The **render** option allows to override or extend the default mark’s rendering method. See [extengin Plot](./extensions.md#render) for details.

Author: ```suggestion
The **render** option allows to override or extend the default mark’s rendering method. See [extending Plot](./extensions.md#render) for details.
```

```

A convenience method for composing a mark from a series of other marks. Returns an array of marks that implements the *mark*.plot function. See the [box mark](../marks/box.md) implementation for an example.

> ## Custom marks

Reviewer: ```suggestion
## Rendering
```

If this section is about custom marks, it should also cover extending the Mark class and implementing other Mark methods such as *mark*.initialize. But that adds scope to this PR, so let’s keep it focused on rendering.

Also, this section describes both **render functions** (functional marks, shorthand for a custom mark) and **render transforms**. I think we should make clear that these are two different concepts even though they are similar.

> Each [mark](./marks.md) has a render method that is called once for each facet (unless its data for that facet is empty). This method is responsible for drawing the mark. You can extend (or replace) this method by specifying the **render** mark option as a custom function.

Reviewer: ```suggestion
A mark’s render method is called once for each facet, unless its data for that facet is empty. The render method is responsible for drawing the mark by producing an SVG element. You can extend or replace this method by specifying the **render** mark option as a custom function.
```

:::warning

> Note that, in general, we do not recommend using this low-level interface when a more high-level option exists. It is meant for use by extension developers more than by users.

Reviewer: ```suggestion
Note that, in general, we do not recommend using this low-level interface when a more high-level option exists such as a [data transform](https://observablehq.com/plot/features/transforms). It is meant for use by extension developers more than by users.
```

:::

The render function is called with the following six arguments:

* *index*: the index of the facet
* *scales*: the scale functions and descriptors
* *values*: the scaled and raw channels
* *dimensions*: the dimensions of the facet
* *context*: the context

> * *next*: the next render method in the chain

Reviewer: The *next* argument only applies to render transforms, not render functions.

> The function is expected to return a single SVG node, or null or undefined if the facet should be skipped. Typically, it returns a [&lt;g>](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/g) container, with a child node (say, a [&lt;circle>](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle)) for each valid data point.

Reviewer: ```suggestion
The function is expected to return a single SVG node, or null or undefined if no output is desired for the current facet. Typically, it returns a [G element](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/g), with a child node (say, a [circle element](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/circle)) for each valid data point.
```

The *index* is an array of indices in the channels, that represent the points to be drawn in the current facet. The *scales* object contains the scale functions, indexed by name, and an additional scales property with the scales descriptors, also indexed by name.

For example, the following code will log the color associated with the Torgersen category ("#e15759") and the [instantiated color scale object](./plots.md#plot_scale), and will not render anything to the chart.

```js
Plot.dot(penguins, {
  x: "culmen_length_mm",
  y: "culmen_depth_mm",
  fill: "island",
  render(index, scales, values, dimensions, context, next) {

>     console.log(scales.color("Torgersen"));

Reviewer: ```suggestion
    console.log(scales.color("Torgersen")); // "#e15759"
```

    console.log(scales.scales.color);
  }
}).plot()
```

The *values* object contains the scaled channels, indexed by name, and an additional channels property with the unscaled channels, also indexed by name. For example:

```js
Plot.dot(penguins, {
  x: "culmen_length_mm",
  y: "culmen_depth_mm",
  fx: "species",
  fill: "island",
  render(index, scales, values, dimensions, context, next) {
    const i = index[0];
    console.log(i, values.fill[i], values.channels.fill.value[i]);
  }
}).plot()
```

will output the following three lines to the console, with each line containing the index of the first penguin of the current facet, its fill color, and the underlying (unscaled) category:

```js
0 '#e15759' 'Torgersen'
152 '#f28e2c' 'Dream'
220 '#4e79a7' 'Biscoe'
```

The *dimensions* object contains the marginTop, marginRight, marginLeft,marginBottom, and width and height of the chart. For example, to draw an ellipse that extends to the edges:

```js
Plot.plot({
  marks: [

>     function (index, scales, values, dimensions, context, next) {

Reviewer: ```suggestion
    function (index, scales, values, dimensions, context) {
```
